rs/consensus/src/consensus/validator.rs

```diff
@@ -956,12 +1025,19 @@ impl Validator {
                 // proposals proceed to be checked normally.
             }
 
+            // Skip block proposals with a disqualified rank. We do this after
+            // checking for a fast-path validation, to avoid getting stuck.
+            if disqualified_ranks
+                .get(proposal.height(), proposal.rank())
+                .is_some()
+            {
+                continue;
+            }
+
             // Skip validation and drop the block if it has a higher rank than a
-            // known valid block. Note that this must happen after we first allow
-            // "block with notarization" validation (see above). Otherwise we may
-            // get stuck when a block maker equivocates.
-            if let Some(Some(min_rank)) = known_qualified_ranks.get(&proposal.height()) {
-                if proposal.rank() > *min_rank {
+            // known valid block.
+            if let Some(min_rank) = valid_qualified_ranks.get_lowest_rank(proposal.height()) {
+                if proposal.rank() > min_rank {
                     // Skip them instead of removal because we don't want to end up
                     // requesting these artifacts again.
```

Comment:
Isn't the main reason also that the proposal with higher rank might still be notarized in the future?

Reply:
Yes. Also the comment above seems incorrect, we're not even dropping the block. I rewrote it, and merged the two comments.